keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2144 @@
+  labels:
+    app: critical-ds
+  name: ae9b4bc6-e4ca-4f8e-962b-2d4459b1f684-critical-ds-5nxjs
+  namespace: critical-workloads
+spec:
+  podRef:
+    name: critical-ds-5nxjs
+    uid:  ae9b4bc6-e4ca-4f8e-962b-2d4459b1f684
+  progressDeadlineSeconds: 1800
+  evacuators:
+    - evacuatorClass: daemonset.apps.k8s.io
+      priority: 10000
+      role: controller
+status:
+  activeEvacuatorClass: daemonset.apps.k8s.io
+  activeEvacuatorCompleted: false
+  evacuationProgressTimestamp: "2024-04-22T21:40:32Z"
+  expectedEvacuationFinishTime: "2024-04-22T21:41:32Z" # now + terminationGracePeriodSeconds:
+  failedEvictionCounter: 0
+  message: "critical-ds is terminating the pod due to node maintenance (OS upgrade)."
+  conditions: []
+```
+
+Once the pod is terminated and removed from the node, it should not be re-scheduled on the node by
+the deamonset controller until the node maintenance is complete.
+
+### Kubelet: Graceful Node Shutdown
+
+The current Graceful Node Shutdown feature has a couple of drawbacks when compared to
+NodeMaintenance:
+- It is application agnostic as it only provides a static grace period before the shutdown based on
+  priority. This does not always give the application enough time to react and can lead to data
+  loss or application availability loss.
+- The DaemonSet pods may be running important services (critical priority) that should be available
+  even during part of the shutdown. The daemon set controller does not have the observability of
+  the kubelet shutdown procedure and cannot infer which DaemonSets should stop running. The
+  controller needs to know which DaemonSets should run on each node with which priorities and
+  reconcile accordingly.
+
+To support these use cases we could introduce a new configuration option to the kubelet called
+`preferNodeMaintenanceDuringGracefulShutdown`.
+
+This would result in the following behavior:
+
+When a shutdown is detected, the kubelet would create a NodeMaintenance object for that node.
+Then it would block the shutdown indefinitely, until all the pods are terminated. The kubelet
+could pass the priorities from the `shutdownGracePeriodByPodPriority` to the NodeMaintenance,
+just without the `shutdownGracePeriodSeconds`. This would give applications a chance to react and
+gracefully leave the node without a timeout. [Pod Selection](#pod-selection) would ensure that user
+workloads are terminated first and critical pods are terminated last. 
+
+By default, all user workloads will be asked to terminate at once. The Evacuation API ensures that
+an evacuator is selected or an eviction API is called. This should result in a fast start of a pod
+termination.NodeMaintenance could then be used even by spot instances.
+
+The NodeMaintenance object could survive kubelet restarts, and the kubelet would always know if the
+node is under shutdown (maintenance). The cluster admin would have to remove the NodeMaintenance
```

Comment:
Would the shutdown cancellation work in the same way as it currently is under Graceful Node Shutdown? I assume, yes, it would.

Reply:
This KEP also supports node maintenance for Windows nodes, where graceful shutdown handling is (AIUI) not implemented.

Reply:
IMO you can mark a node as needing maintenance even if it doesn't run systemd (either Linux without systemd, or Windows which doesn't really have an init process in the POSIX sense). So this KEP can move forward whether or not your nodes gracefully handle a shutdown signal.

Reply:
[Cancelation by kubelet](https://kubernetes.io/docs/concepts/cluster-administration/node-shutdown/#graceful-node-shutdown)
>  Note:
There are cases when Node termination was cancelled by the system (or perhaps manually by an administrator). In either of those situations the Node will return to the Ready state. However, Pods which already started the process of termination will not be restored by kubelet and will need to be re-scheduled.

Is the node termination cancellation reliable? Can we use it for deletion of the NodeMaintenance?

---

Also the behavior is application specific, when the the NodeMaintenance is cancelled/deleted (please see the [Evacuation KEP](https://github.com/kubernetes/enhancements/pull/4565)). The application/evacuator can  terminate the pod or it can also recover.